src/crafting.cpp

```diff
@@ -24,6 +24,28 @@
 #include <math.h>    //sqrt
 #include <algorithm> //std::min
 
+enum usage {
+	use_from_map = 1,
+	use_from_player = 2,
+	use_from_both = 1 | 2,
+	use_from_none = 4,
+	cancel = 5
+};
+
+struct comp_selection {
```

Comment:
For what would you need constrains here? Usually, the compiler complains if the type for whatever reason doesn't work (doesn't fit the constrains):

```C++
// note: CompType needs a foo() function.
template<typename CompType>
struct comp_selection {
    usage use_from;
    CompType comp;

    void bar() {
        comp.foo(); // compiler complains here if the specific CompType doesn't have foo.
    }
};
```

That's not really optimal, but it seems good enough with the (very limited) use here.